boot: EFI mixed mode support

medhefgo opened the pull request:
This is only tested on QEMU so far and I would prefer to merge this only after it is confirmed that it works on real hardware. Note that OVMF has its own COFF emulation support, so to properly test this a suitable OVMF build without support for it has to be used.

I considered making ia32 builds on 64bit machines automatic, but that would probably make the whole meson logic for gnu-efi even more unsightly. I think asking distros to use a second build dir for this is reasonable. If desired, a "systemd-boot" target alias could be provided so simplify building a bit.

The bootctl changes are only compile-tested.

poettering commented on src/boot/efi/util.h:
```diff
@@ -24,6 +24,17 @@
+#  define FMT_D L"%ld"
+#else
+#  error "Unexpected pointer size"
+#endif
```
hmm, so C99 PRIu32-style macros do not include the `%` prefix. Our FMT_UID, … macros generally do. That has a reason: the PRIu32-style macros are generic, and still allow you to specify alignment, precision and so on. THe FMT_UID macros on the other hand don't allow that, because we now pretty excactly how UIDs should be formatted.

Now, this here is a generic use case, it is a way to display any natively sized integer. Hence it probably should not contain the "%" prefix.

Hence maybe do:

```c
#define PRIuN L"u"
```

and similar instead, and then use it as:

```c
SPrint(…, "foo %" PRIuN " bar");
```

pretty? no! but I think the most obvious way to do this for people who have accepted C99 syle PRI* macros into their lives.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -138,24 +137,20 @@ static void cursor_right(
         line = xnew(CHAR16, size);
         print = xnew(CHAR16, x_max + 1);
-        StrCpy(line, line_in);
+        StrCpy(line, *line_in ?: L"");
```
we probably should move strempty() into the fundamental stuff shared between userspace and efispace

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2239,6 +2224,36 @@ static void config_load_xbootldr(
+        assert(entry);
+        assert(ret_options);
+
+        if (entry->type != LOADER_LINUX || !entry->initrd)
```
why the entry-type conditionalization? i.e. the right hand conditionalization should conver that too? i.e. if the type is not LOADER_LINUX then initrd is gonna be NULL anyway, no?

medhefgo replied:
You can create an LOADER_EFI config entry with initrds just fine. Does it make any particular sense? Not really. Also, this is just defensive programming.

poettering replied:
good point

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2251,6 +2291,8 @@ static EFI_STATUS initrd_prepare(const ConfigEntry *entry, CHAR16 **ret_options)
 
         *ret_options = TAKE_PTR(options);
+        *ret_initrd = TAKE_PTR(initrd);
+        *ret_initrd_size = size;
```
the upper EFI_SUCESS code paths should initialize these to NULL/0, as per coding style

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,46 @@ static void locate_sections(
+                    compat->machine_type == TARGET_MACHINE_TYPE)
+                        return compat->entry_point;
+
+                addr += compat->size;
```
overflow checks?

medhefgo replied:
The `compat->size > size` some lines above should take care of this, no?

poettering commented on src/boot/efi/pe.c:
```diff
@@ -171,20 +217,30 @@ EFI_STATUS pe_alignment_info(
 
         pe = (const struct PeFileHeader*) ((const UINT8 *)base + dos->ExeHeader);
-        if (!verify_pe(pe))
+        if (!verify_pe(pe, TRUE))
```
maybe write as:

```c
if (!verify_pe(pe, /* allow_alt= */ TRUE))
```

poettering commented on src/boot/bootctl.c:
```diff
@@ -251,6 +259,30 @@ static int get_file_version(int fd, char **v) {
+        if (streq(platform_size, "32"))
+                return EFI_MACHINE_TYPE_ALT_NAME;
+
+        log_warning("Unknown EFI platform size: %s", platform_size);
```
for warnings we usually say what we are doing now, i.e. suffix with ", ignoring" or so. But in this case we should probably say ", using the default word size"

poettering commented on src/boot/bootctl.c:
```diff
@@ -251,6 +259,30 @@ static int get_file_version(int fd, char **v) {
+        if (r == -ENOENT)
+                return EFI_MACHINE_TYPE_NAME;
+        if (r < 0) {
+                log_error_errno(r, "Error reading EFI firmware platform size: %m");
```
given this is not fatal for anything, it's mostly a warning and thus should be downgraded to log_warning_errno()...

So a "platform size" is a weird concept. I figure they mean "word size"? maybe write that in the log message, i.e.

"Error reading EFI firmware platform word size, using the default word size"

poettering commented on man/bootctl.xml:
```diff
@@ -277,6 +277,17 @@
+        <para>Defaults to <literal>auto</literal>, which will use the kernel-reported EFI platform size of
+        the running system to decide the right binaries to use. Setting this to <literal>all</literal> will
+        install binaries for all available architectures (this implies <option>--no-variables</option>).</para></listitem>
+      </varlistentry>
```
so this feature as it looks right now seems to support exactly one or two archs, which are compiled into bootctl. 

I think we should make this entirely generic, so that we copy basically `/usr/lib/systemd/boot/efi/systemd-boot*.efi`, i.e. any arch that is supported. Or if you specifiy --arch= any subset of it.

i.e. I think distros should be able to implement multiarch to the max if they like and have arm and x86 EFI binaries of sd-boot in place if they like, and "bootctl" should be able to drop it in.

Or in otherwords: instead of compiled in list of archs, just pick up every file that is installed.

medhefgo replied:
> Or in otherwords: instead of compiled in list of archs, just pick up every file that is installed.

This is what the `all` arch is for. And I believe this should be explicitly opt-in. I certainly wouldn't want bootctl to always install ia32 binaries when I don't need them.

I now made it so that you can select from any efi arches we support, but we still default to auto detecting and only installing for that arch, which is imho the right thing to do.

poettering replied:
I'd argue here that if you have those other flavours installed, then we should install them too. i.e. I'd see that as a drop-in thing: if the other flavours are installed we take them too. Why would you install them in /usr/ if you wouldn't also want them in /efi/?

poettering commented on src/boot/efi/meson.build:
```diff
@@ -10,17 +10,26 @@ if not get_option('efi') or get_option('gnu-efi') == 'false'
         subdir_done()
 endif
 
```
> But arch for example has no 32bit port anymore. There you'd want to have a way to build ia32 without having to do a full 32bit cross build just so you can toss out most of it. You'd need to provide a minimum of a lot of 32bit dependencies just so meson could run successfully, even when you just want to issue the build commands for the ia32 efi binaries to speed things up.

But that sounds like a workaround on a deliberate choice that arch made, that they don't want to support 32bit anymore. Shoehoring a 32bit build into a pretend-64bit build is a hack, no?

anyway, @keszybz what's your take on this?

medhefgo replied:
It is perfectly fine for a distro to drop 32bit kernel/userspace. But that still means that you need to provide the necessary tools to boot your 64bit OS. And there are 64bit systems that have been crippled with a 32bit EFI firmware and arch should be able to easily support this. (I feel by your logic it should drop i386-pc grub too.)

Also, I highly doubt dpkg even allows cross-arch package installs (lib32-* packages are still amd64 packages). What seem to suggest is pulling ia32 binaries out of the i386 dpkg and stuff it into the amd64 dpkg. That just sounds like a terrible idea.

@bluca you seem to be involved in debian, what's your take on this?

poettering replied:
So yeah, what @bluca describes how Ubuntu does it makes sense to me: drop i386 mostly, but do i386 builds still for a select set of packages.

xnox replied:
Debian and Ubuntu both have multiarch support and allow installing foo:i386 package on an amd64 system for example. (which results in installing or coinstalling foo_1.23-3_i386.deb and foo_1.23-3_amd64.deb on a single system). However we have been doing some interesting packaging to build package named foo-efi-ia32 as either _amd64 or _all whilst actually containing the EFI 32-bit binaries. So building/installing stuff in general has solutions.

In terms of hardware, I know that obsolete Macbooks Intel Core Duo are 64bit CPU but 32bit efi (note, not Core 2 Duo). And there were Intel MinnowBoards that had 64-bit cpu, had both 32 bit & 64 bit EFI firmware available, however by default out of the factory they shipped with the 32 bit firmware.

I am not aware of any other popular / large systems that have 32bit EFI firmware yet 64bit CPU. My understanding is that no other SKU/vendor repeated that mistake since. And all platforms since ship with matching EFI firmware and CPU, being 64-bit one.

Minnow board users should reflash into 64bit EFI firmware. The core duo macbook I had in University, it is long obsolete and not supported in terms of power efficiency CPU disk space etc. They can continue to use whatever they have on them, and shouldn't need to support or install new OS on them. Their batteries should be long dead too, and impossible to replace.

I am interested to know if there are any other SKUs that are being manufactured and if anybody supports them that ship ia32 efi firmware. I.e. stuff released in the last 10 years, i.e. from 2013 onwards. And has received Spectre Meltdown mitigations.

xnox replied:
Yeap, after my post above I got pointed at machines sold in 2015 with such setup. Not flagships or anything, but yeah, they were still made and sold less than 8 years ago.

mirh replied:
Code review is def not the place for technical support, let alone when you bring in problems that are obviously distro-specific.

Viking8 replied:
@mirh I didn't ask for any tech support. I was just relaying some info about a device that uses a 64-bit CPU and a 32-bit UEFI. The rest was a little back story in case someone was still working on this so they would know a little about what I did.

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,46 @@ static void locate_sections(
 
+static UINT32 get_compat_entry_address(const struct DosFileHeader *dos, const struct PeFileHeader *pe) {
+        UINTN addr = 0, size = 0;
+        const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
```
static?

poettering commented:
so in the context of multib we usually talk about the "secondary" arch, which appears to be conceptually close to what you refer to here as "alt" arch. Is the "alt" moniker something used by EFI specs anywhere? Otherwise I#d call this "secondary" here too, to minimize the concepts we introduce here.

I'd prefer if the build system stuff to allow building a 64bit systemd with a 32bit sd-boot would be split out into a separate arch, it seems to be something only some distros are looking for, and not sure it should go in this way.

xnox commented:
From Ubuntu's point of view, we have never fully provided ia32 EFI enabled installer media or support. And ideally we wouldn't want for that to be built by default during x86_64 build. It is best for this to become an opt-in feature.

medhefgo commented:
> so in the context of multib we usually talk about the "secondary" arch, which appears to be conceptually close to what you refer to here as "alt" arch. Is the "alt" moniker something used by EFI specs anywhere? Otherwise I#d call this "secondary" here too, to minimize the concepts we introduce here.

The efi spec says nothing here, as this is out of scope. It really doesn't matter to me what name we use, but considering the kernel puts this in a ".compat" section maybe use that?

> I'd prefer if the build system stuff to allow building a 64bit systemd with a 32bit sd-boot would be split out into a separate arch, it seems to be something only some distros are looking for, and not sure it should go in this way.

What you mean separate arch? Also, what's the big deal about it?

It's a very small change as it's mostly just passing the efi-cflags to a bunch of calls we make. You can hardly call this hacky/shoehorning imho and has real benefits/use cases.
In fact, it eases the ability to do test/debug builds for ia32 (as we both know, the ci has caught the occasional case due to 32 vs 64 UINTN), so that you don't have to setup a full blown qemu container for this or wait for ci to complain.

Also, all of this is clearly opt-in. Unless you *explicitly* ask for it, you will get your regular x64 build.

poettering commented:
> The efi spec says nothing here, as this is out of scope. It really doesn't matter to me what name we use, but considering the kernel puts this in a ".compat" section maybe use that?

Sounds better than "alt" to me. But not a fan of needlessy cryptic abbreviations. I'd probably still call it secondary then. or at least "compatibility"...

poettering commented:
> What you mean separate arch? Also, what's the big deal about it?

Building a cross-binary from an otherwise same-arch tree sounds really weird, and a specific ack around one distro which killed one arch but actually still wants to support it at one place...

> It's a very small change as it's mostly just passing the efi-cflags to a bunch of calls we make. You can hardly call this hacky/shoehorning imho and has real benefits/use cases. In fact, it eases the ability to do test/debug builds for ia32 (as we both know, the ci has caught the occasional case due to 32 vs 64 UINTN), so that you don't have to setup a full blown qemu container for this or wait for ci to complain.

Well, but I might instead just cross build the whole of systemd for i386, no? For the testing usecase that'd be a ton more obvious and more enlightening. i.e. why should the UEFI stuff be so special there that we'd support builds that officially are for arch X but some bits are actually for arch Y.

poettering commented on src/boot/efi/util.c:
```diff
@@ -57,7 +57,7 @@ EFI_STATUS efivar_set_uint_string(const EFI_GUID *vendor, const CHAR16 *name, UI
         assert(name);
 
-        SPrint(str, ELEMENTSOF(str), L"%u", i);
+        SPrint(str, ELEMENTSOF(str), "%" PRIuN, i);
```
where did the L go?

medhefgo replied:
Into the ether? :D

medhefgo commented:
> Building a cross-binary from an otherwise same-arch tree sounds really weird, and a specific ack around one distro which killed one arch but actually still wants to support it at one place...

That's not what arch is doing. It dropped i386 kernel/userspace. But you *still need tools* to get there. For non-EFI boot that's i386 bootloaders. For *crippled* ia32 EFIs on x86_64 CPUs, you need a *ia32* EFI bootloader.

> Well, but I might instead just cross build the whole of systemd for i386, no? For the testing usecase that'd be a ton more obvious and more enlightening. i.e. why should the UEFI stuff be so special there that we'd support builds that officially are for arch X but some bits are actually for arch Y.

But there is no point in doing that. You literally only need ia32 EFI binaries and that's it. No bootctl, no lib32 systemd, nada, nothing else. Meanwhile, to get this working, you'd have to pull in a ton of lib32 stuff (which has to be build and maintained). Just so you drop it all after the build because all you are interestd in are ia32 efi binaries.

poettering commented:
I'll leave the cross-build thing for @keszybz to okay, as our resident meson guru.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -531,6 +534,8 @@ static void print_status(Config *config, CHAR16 *loaded_image_path) {
                     Print(L"        device: %D\n", DevicePathFromHandle(entry->device));
                 ps_string(L"        loader: %s\n", entry->loader);
+                STRV_FOREACH(initrd, entry->initrd)
+                    Print(L"        initrd: %s\n", *initrd);
```
that indentation is killing me

medhefgo commented:
> I'll leave the cross-build thing for @keszybz to okay, as our resident meson guru.

You keep saying that like a mantra. But repeating it is not gonna improve the meson bus-factor… :joy_cat:

poettering commented on src/boot/efi/boot.c:
```diff
@@ -1454,18 +1459,12 @@ static void config_entry_add_from_file(
-
+                        entry->initrd = xreallocate_pool(
+                                entry->initrd,
+                                n_initrd * sizeof(UINT16 *),
```
the old size is `(n_initrd+1)*sizef(…)`, no?

bluca commented:
> > Building a cross-binary from an otherwise same-arch tree sounds really weird, and a specific ack around one distro which killed one arch but actually still wants to support it at one place...
> 
> That's not what arch is doing. It dropped i386 kernel/userspace. But you _still need tools_ to get there. For non-EFI boot that's i386 bootloaders. For _crippled_ ia32 EFIs on x86_64 CPUs, you need a _ia32_ EFI bootloader.
> 
> > Well, but I might instead just cross build the whole of systemd for i386, no? For the testing usecase that'd be a ton more obvious and more enlightening. i.e. why should the UEFI stuff be so special there that we'd support builds that officially are for arch X but some bits are actually for arch Y.
> 
> But there is no point in doing that. You literally only need ia32 EFI binaries and that's it. No bootctl, no lib32 systemd, nada, nothing else. Meanwhile, to get this working, you'd have to pull in a ton of lib32 stuff (which has to be build and maintained). Just so you drop it all after the build because all you are interestd in are ia32 efi binaries.

This makes sense to me - while Debian derivatives are not there _yet_, at some point that's the direction everyone has to move toward, simply because the cost of maintaining an entire bootstrappable toolchain just for an EFI bootloader would be so high. Of course 32bit EFI firmwares might completely disappear before we get to that point, which I dearly hope for, but if the latest hardware is from 2015 it's not _that_ long ago, sadly.
And for non-Debian derivatives, without full multiarch, this makes sense already today. The alternative would be to add enough meson glue that you can build _only_ the EFI binaries on a native build, but my spider sense tells me that would be way more invasive.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2344,14 +2344,19 @@ static EFI_STATUS image_start(
 
+        EFI_LOADED_IMAGE *loaded_image;
+        err = BS->OpenProtocol(
+                image,
```
we actually usually indent continuation after `(` to 16chars, not 8

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2361,14 +2366,31 @@ static EFI_STATUS image_start(
                  * (userspace) program. */
-                graphics_mode(FALSE);
-                return log_error_status_stall(err, L"Failed to execute %s (%s): %r", entry->title_show, entry->loader, err);
+                return err;
```
doesn't matter, but maybe `return EFI_SUCCESS` here is more obvious to read?

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2324,12 +2324,47 @@ static EFI_STATUS image_start(
+        if (err == EFI_UNSUPPORTED && entry->type == LOADER_LINUX) {
+                /* If LoadImage fails, we may still start the kernel manually the same way we do for the
+                 * unified stub loader, which will then call the compat entry point if one exists. *
+
```
spurious empty line?

poettering commented on src/boot/bootctl.c:
```diff
@@ -780,10 +809,17 @@ static int install_binaries(const char *esp_path, bool force) {
+                const char *arch = arg_arch ?: efi_arch();
+                suffix = strjoina(arch, suffix);
+                suffix_signed = strjoina(arch, suffix_signed);
+        }
```
btw, i think we should generate an error if the selected arch is not supported, i.e. no binary installed for this arch

medhefgo replied:
This should probably be done separately, as it currently doesn't warn if nothing was copied.

poettering commented on src/boot/bootctl.c:
```diff
@@ -1313,6 +1349,7 @@ static int help(int argc, char *argv[], void *userdata) {
                "                       variables cannot be written\n"
                "     --make-machine-id-directory=yes|no|auto\n"
                "                       Create $BOOT/$MACHINE_ID\n"
+               "  -a --arch=ARCH       EFI arch for installed binaries\n"
```
abbreviation seems unnecessary. I mean you have it a short option "a" already, no need to abbreviate the long option.

poettering commented on src/boot/bootctl.c:
```diff
@@ -1416,6 +1454,17 @@ static int parse_argv(int argc, char *argv[]) {
+                                        return log_oom();
+                        } else
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Unknown EFI arch: %s", optarg);
+                        break;
```
i have the suspicion this should actually permit being used multiple times, i.e. be an strv, not a simple string.

And I still think that if unspecified we should copy boot loaders for all archs for which one is installed. i.e. if people don't want some arch they shouldn't install the package for it, or conversely: installing the package for an arch should make the boot lodaer just work for that arch.

poettering commented on src/boot/bootctl.c:
```diff
@@ -1924,11 +1974,23 @@ static int verb_install(int argc, char *argv[], void *userdata) {
+        if (!arg_touch_variables)
+                return r;
+
+        if (streq_ptr(arg_arch, "all"))
```
allowin an arch that is magic is evil... at the very least it should be  macro, so that it's obvious it's special. But actually I think there should be a separate boolean for the special cases. No need to shove all special cases into a single string.

medhefgo commented:
(I dropped the `linux_exec` after `LoadImage` as it doesn't seem to work for the one user that I added it for (it was a long shot).)

For now I made bootctl simply default to auto-detection with a `--all-architectures` option (with no specific arch selection option). Partly because I am not particularly invested into bootctl and wanted to keep things simple.

I also disagree that we should default to installing every arch by default as probably 99% of use cases will be "install this on the currently running system" where we can simply detect the arch anyways and would want to change the efi vars too (which pretty much has to be arch dependent).
Also, not every distro splits its packages into 100+ subpackages just for the sake of it (like debian for example). And simply installing everything that is available when only one arch can be used anyways is just wasteful to me. And the detection is quick and easy to do.

poettering commented:
needs rebase

poettering commented on src/boot/efi/assert.c:
```diff
@@ -6,7 +6,7 @@
 
 void efi_assert(const char *expr, const char *file, unsigned line, const char *function) {
-        log_error_stall(L"systemd-boot assertion '%a' failed at %a:%u, function %a(). Halting.", expr, file, line, function);
+        log_error_stall(L"systemd-boot assertion '%a' failed at %a:%" PRIuN L", function %a(). Halting.", expr, file, line, function);
```
is this right?

reading the comments in `PrintLib.h` it appears to me %u is size of an int, and hence also size of an unsigned int, and hence the code actually was right here

poettering commented on src/boot/efi/boot.c:
```diff
@@ -477,7 +477,7 @@ static void print_status(Config *config, CHAR16 *loaded_image_path) {
           ps_bool(L"                  shim: %s\n",      shim_loaded());
           ps_bool(L"                   TPM: %s\n",      tpm_present());
-            Print(L"          console mode: %d/%d (%lux%lu @%ux%u)\n", ST->ConOut->Mode->Mode, ST->ConOut->Mode->MaxMode - 1LL, x_max, y_max, screen_width, screen_height);
+            Print(L"          console mode: %d/%ld (%" PRIuN L"x%" PRIuN L" @%ux%u)\n", ST->ConOut->Mode->Mode, ST->ConOut->Mode->MaxMode - INT64_C(1), x_max, y_max, screen_width, screen_height);
```
I'd really prefer if we'd use PRIu32 style stuff for fixed size integers, like `screen_width` and `screen_height`. then it's much cloer to how we do stuff in userspace

medhefgo replied:
When I introduced this line you got paranoid about potential underflows should a crazy firmware give us a INT32_MIN as MaxMode. This is just continuing to cover our asses (and I didn't know at the time that `_C` macros existed or I would've used that). :P

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2570,11 +2571,10 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
+                        /* Not using EFI_ERROR here because positive values are also errors like with any
+                         * other (userspace) program. */
                         goto out;
-                }
```
seems like an unrelated change? where di the graphics mode thing go?

poettering commented on src/boot/efi/boot.c:
```diff
@@ -138,24 +137,20 @@ static void cursor_right(
-                line_in = L"";
-
-        len = StrLen(line_in);
+        len = *line_in ? StrLen(*line_in) : 0;
```
btw, in userspace we have `strlen_ptr` for this. mabe add here too?

poettering commented on src/boot/efi/boot.c:
```diff
@@ -319,8 +314,10 @@ static BOOLEAN line_edit(
-                                *line_out = TAKE_PTR(line);
+                        if (!streq(line, *line_in)) {
+                                FreePool(*line_in);
+                                *line_in = TAKE_PTR(line);
```
btw, in userspce we have free_and_replace() for this

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2251,28 +2251,64 @@ static void config_load_xbootldr(
         }
 
+        /* Note that order of initrds matters. The kernel will only look for microcode updates in the very
+         * first one it sees. */
```
not just for this, but later initrds override contents of earlier ones, if they contain same inodes

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2367,14 +2367,19 @@ static EFI_STATUS image_start(
         if (EFI_ERROR(err))
                 return log_error_status_stall(err, L"Error registering initrd: %r", err);
 
+        EFI_LOADED_IMAGE *loaded_image;
```
shouldn't we issue CloseProtocol in error paths on this?

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,45 @@ static void locate_sections(
+static UINT32 get_compatibility_entry_address(const struct DosFileHeader *dos, const struct PeFileHeader *pe) {
+        UINTN addr = 0, size = 0;
+        static const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
+
```
a comment here what this does would be great, as this is totally not obvious for the drive-by reader

poettering replied:
i mean the function as a whole, not this specific line here

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,45 @@ static void locate_sections(
+        UINTN addr = 0, size = 0;
+        static const CHAR8 *sections[] = { (CHAR8 *) ".compat", NULL };
+
+        locate_sections((struct PeSectionHeader *) ((UINT8 *) dos + section_table_offset(dos, pe)),
```
doesn't matter but strictly speaking you cast the const away, that `dos` carries. to be fully correct the two casts would still have it

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,45 @@ static void locate_sections(
+                UINT8 size;
+                UINT16 machine_type;
+                UINT32 entry_point;
+        } _packed_ LinuxPeCompat1;
```
is the `_packed_` about actual packing, or about alignment?

poettering commented:
did another round of reviews. mostly looks good. not sure about the multlib build thing. will defer to @keszybz on that one, given it#s a build system issue, and he's more immersed in that thatn I am.

poettering commented:
any chance you can split out the build system thing, into a separate PR, so that we can discuss this separately? the other stuff looks good to be merged quickly now.

medhefgo commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,45 @@ static void locate_sections(
+                        return compat->entry_point;
+
+                addr += compat->size;
+                size -= compat->size;
```
From what I've learned recently with the bcd test code was that the compiler seems to do the right thing when you access a (packed?) struct that's misaligned. At least I didn't get ubsan warnings for the structs, while accessing raw primitive types that way did.

Btw, the PE parsing code as a whole isn't really on the level of paranoid that it should be…

medhefgo replied:
I was saying that despite x86 cpus supporting unaligned access, ubsan was able to detect those. So I am assuming that it's a non-issue for (packed) structs; the compiler does some magic there.

Nevertheless, having an explicit check doesn't hurt.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2438,10 +2437,29 @@ static EFI_STATUS image_start(
         efivar_set_time_usec(LOADER_GUID, L"LoaderTimeExecUSec", 0);
         err = BS->StartImage(image, NULL, NULL);
         graphics_mode(FALSE);
```
this loks misplaced. if we use the compat proto next, then we should only switch from graphics mode once that fails, too, no?

medhefgo replied:
This is deliberate to keep things simple. And it shouldn't really matter anyways.

poettering commented on src/boot/bootctl.c:
```diff
@@ -2002,11 +2047,14 @@ static int verb_install(int argc, char *argv[], void *userdata) {
+                return log_warning_errno(r, "Not changing EFI variables with --all-architectures.");
+
+        char *path = strjoina("/EFI/systemd/systemd-boot", arch, ".efi");
+        r = install_variables(arg_esp_path, part, pstart, psize, uuid, path, install);
```
just do `return install_variables(…);`

poettering commented on src/boot/bootctl.c:
```diff
@@ -2002,11 +2047,14 @@ static int verb_install(int argc, char *argv[], void *userdata) {
-                                      "/EFI/systemd/systemd-boot" EFI_MACHINE_TYPE_NAME ".efi",
-                                      install);
+        if (!arg_touch_variables)
+                return r;
```
return 0;

poettering commented on src/boot/bootctl.c:
```diff
@@ -2002,11 +2047,14 @@ static int verb_install(int argc, char *argv[], void *userdata) {
+                return r;
+
+        if (arg_arch_all)
+                return log_warning_errno(r, "Not changing EFI variables with --all-architectures.");
```
you are propagating `r` here, which has no useful value. You want to return 0 here. 

also, LOG_WARN is too loud of a log level. It's LOG_NOTICE or LOG_INFO at best, since it's nothing actionable at all.

i.e.:

```c
if (arg_arch_all) {
        log_info(…);
        return 0;
}
``

medhefgo commented:
I've done away with the format macros in favor of just casting. The `Print()` API is just a mess and since you technically need to deal with `int/unsigned/(U)INT{32,64}` differently, casting seems so much nicer to me. Also, I hate int promotion. Lemme know what you think.

medhefgo commented:
(if you haven't noticed, I switched it back to PRI macros)

poettering commented on src/boot/efi/util.c:
```diff
@@ -57,7 +57,7 @@ EFI_STATUS efivar_set_uint_string(const EFI_GUID *vendor, const CHAR16 *name, UI
         assert(name);
 
-        SPrint(str, ELEMENTSOF(str), L"%u", i);
+        ValueToString(str, FALSE, i);
```
hmm, this looks wrong. ValueToString() takes a signed integer... but this one is about unsigned

medhefgo replied:
This is exactly what `SPrint` does internally. It does not give a fuck about the sign and will pass UINT32/64 to the INT64 arg of `ValueToString`. Which is why `Print(L"%lu", UINTN_MAX)` gives you `-1`.

I know this is messed up. But I think it's better to call this directly rather than indirectly through the Print machinery.

poettering replied:
please add inline comment about this.

poettering commented on src/boot/efi/boot.c:
```diff
@@ -2301,27 +2301,63 @@ static void config_load_xbootldr(
                 const ConfigEntry *entry,
-                CHAR16 **ret_options) {
+                CHAR16 **ret_options,
+                UINT8 **ret_initrd,
```
VOID**?

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,52 @@ static void locate_sections(
+        } _packed_ LinuxPeCompat1;
+
+        while (size >= sizeof(LinuxPeCompat1)) {
+                LinuxPeCompat1 *compat = (LinuxPeCompat1 *) ((UINT8 *) dos + addr);
```
according to C specs you aren't even allowed to to *take* an unaligned pointer of a type, or you are already in undefined behaviour territory. (i.e. the compiler may assume that a pointer of type LinuxPeCompat* is always aligned, and hence the check below is something it can optimize away, since it already knows it is, it after all has the right type).



i.e. you need to check for alignment *before* you cast to the struct, not after.

poettering commented:
looks good, just some minor stuff left. 

(sorry for the slow review!)

poettering commented on src/boot/efi/pe.c:
```diff
@@ -160,6 +166,49 @@ static void locate_sections(
+                UINT32 entry_point;
+        } _packed_ LinuxPeCompat1;
+
+        while (size >= sizeof(LinuxPeCompat1) && addr % __alignof__(LinuxPeCompat1 *) == 0) {
```
actually, must be `__alignof__(LinuxPeCompat1)`, i.e. of the struct, not the pointer.

might make sense to wrap is in a helper macro, something like this:

```c
#define POINTER_IS_ALIGNED(p, t) ({ \
        const void *_q = (p); \
        UINTN a = (UINTN) _q; \
        (a % __alignof__(t)) == 0; \
})
```

With that you could just do:

```c
while (size >= sizeof(LinuxPeCompat1) && POINTER_IS_ALIGNED((UINT8*) dos + addr, LinuxPeCompat1) {
…
``

medhefgo replied:
Huh. Interestingly, alignof the _packed_ struct is 1, while 4 without. Does checking really make sense then??

poettering replied:
Yeah, the _packed_ removes the alignment requirements, that's why I asked about that earlier.

But I'd still do the check, since in general we should check alignment before assuming too much from arbitrary pointers.

poettering replied:
alternatively, add an `assert_cc(__alignof__(LinuxPeCompat1) == 1)` before casting into a pointer of that type. That documents clearly that the assumption here is that any pointer is safe for this structure.

medhefgo commented:
> lgtm, but let's wait for the next cycle, I feel a bit uncomfortable missing something that changes the boot path so much this late in the cycle given we have no testsuite for sd-boot right now... hope that makes sense?

I don't mind, but I think at least the format specifier and devtree commits should go in now as they're fixing bugs (that nobody has complained about yet). Should I make a PR for them or you gonna cherry pick them?

poettering commented:
hmm, cherry pick isn't straight-forward unfortunately, patches don't apply on git main.

Would be thankful for a split-out PR of those commits. Thanks!

josephlr commented:
@medhefgo I was going to do some TPM2 related testing w/ this. Could you rebase onto latest `main` now that #23073 is merged? I was able to do the rebase locally, but it didn't apply cleanly.

bluca commented:
This is a couple of weeks old and there were a few build changes for sd-boot, so I've rebased to make it go through up-to-date CI, just in case
